File: src/test/programmers/java/level1/DartGame.java
package test.programmers.java.level1;

public class DartGame {
    public static void main(String[] args) {
        System.out.println(solution("1S2D*3T"));
    }

    static int solution(String dartResult) {
        int answer = 0;
        char[] chars = dartResult.toCharArray();
        int[] nums = new int[3];
        int count = -1; //keyPoint

        for(int i = 0; i < chars.length; i++) {
            char ch = chars[i];

            if(ch >= '0' && ch <= '9') {
                count++;
                if(ch == '1' && chars[i + 1] == '0') {
                    nums[count] = 10;
                    i++; //10 으로 인덱스를 두자리 차지하기 때문에 i++
                } else {
                    nums[count] = ch - '0';
                }
            } else if(ch == 'S') {
                nums[count] = (int) Math.pow(nums[count], 1);
            } else if(ch == 'D') {
                nums[count] = (int) Math.pow(nums[count], 2);
            } else if(ch == 'T') {
                nums[count] = (int) Math.pow(nums[count], 3);
            } else if(ch == '*') {
                if(count > 0) { //첫번째 숫자가 아니면 현재 숫자와 전의 숫자 둘다 *2
                    nums[count - 1] *= 2;
                }
                nums[count] *= 2;
            } else if(ch == '#') {
                nums[count] *= -1;
            } else {
                continue;
            }
        }

        answer += nums[0] + nums[1] + nums[2];

        return answer;
    }
}

File: src/test/programmers/python/level1/DartGame.py
def solution(dartResult):
    temp = ''
    score = []
    for i in list(dartResult):
        i = str(i)
        if i.isnumeric():
            temp += i
        elif i == "S":
            temp = pow(int(temp), 1)
            score.append(temp)
            temp = ''
        elif i == "D":
            temp = pow(int(temp), 2)
            score.append(temp)
            temp = ''
        elif i == "T":
            temp = pow(int(temp), 3)
            score.append(temp)
            temp = ''
        elif i == "*":
            if len(score) == 1:
                score[0] *= 2
            else:
                idx = len(score)
                score[idx - 2], score[idx - 1] = score[idx - 2] * 2, score[idx - 1] * 2
        elif i == "#":
            score[-1] *= -1

    return sum(score)

